fuel/datasets/hdf5.py

```diff
@@ -0,0 +1,62 @@
+import tables
+
+from fuel.datasets import Dataset
+
+
+class Hdf5Dataset(Dataset):
+    """An HDF5 dataset
+
+    Parameters
+    ----------
+    sources : tuple of strings
+        Sources which the dataset returns
+    start : int
+        Start index
+    stop : int
+        Stop index
+    data_node : str
+        Parent data node in HDF5 file
+    sources_in_file : tuple of strings
+        Names of nodes in HDF5 file which contain sources. Should the same
+        length as `sources`.
+        Optional, if not set will be equal to `sources`.
+    """
+    def __init__(self, sources, start, stop, path, data_node='Data',
+                 sources_in_file=None):
+        if sources_in_file is None:
+            sources_in_file = sources
+        self.sources_in_file = sources_in_file
+        self.provides_sources = sources
+        self.path = path
+        self.data_node = data_node
+        self.start = start
+        self.stop = stop
+        self.num_examples = self.stop - self.start
+        self.nodes = self._open_file()
+        super(Hdf5Dataset, self).__init__(self.provides_sources)
+
+    def _open_file(self):
+        h5file = tables.openFile(self.path, mode="r")
+        node = h5file.getNode('/', self.data_node)
+
+        return [getattr(node, source) for source in self.sources_in_file]
+
+    def get_data(self, state=None, request=None):
+        """ Returns data from HDF5 dataset.
+
+        .. note:: The best performance if `request` is a slice.
+        """
+        data = [node[request] for node in self.nodes]
+        return data
+
+    def __getstate__(self):
+        fields = self.__dict__
+        # Do not return `nodes` because they are not pickable
+        del fields['nodes']
+        return fields
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        # Open HDF5 file again and sets up `nodes`.
+        self.nodes = self._open_file()
```

Comment:
The `load` method would actually not have the data path hard coded, it would be read from another attribute, which the user could set. So there is no need to overwrite `load`, you would simply change the attribute.

```python
class Bar(object):
    path = '/home/data.txt'

    def load(self):
        self.foo = open(self.path)

    @property
    def foo(self):
        if not hasattr(self, '_foo'):
            self.load()
        return self._foo

    @foo.setter
    def foo(self, value):
        self._foo = value

    def __getstate__(self):
        state = self.__dict__.copy()
        del state['_foo']
        return state

    def baz(self):
         return self.foo.readline()
```
And if you wanted to unpickle this and the path had changed you would just change `bar.path = `/home/new_location/data.txt'`.

We need to factor out the path logic at one point though, or at least standardize it, because it will always be the same: By default, most datasets should read the data from `FUEL_DATA_PATH/dataset_name/*`. Most of the time, when data moves, you can just change `FUEL_DATA_PATH`. However, if the files for some reason aren't in a folder, or the filenames themselves have changed, you need to intervene programmatically I guess. I'm not sure about making this an optional function call though. I'd prefer a method where the folder and filenames are saved as attributes, which can simply be changed e.g.

```python
@do_not_pickle_attributes('images', 'labels')
class MNIST(FileDataset):
    folder = 'mnist'
    files = {
        'train': {'images': 'train-images-idx3-ubyte', 'labels': 'train-labels-idx1-ubyte'},
        'test': {'images': 't10k-images-idx3-ubyte', 'labels': 't10k-labels-idx1-ubyte'}
    }

    def load_attributes(self):
        self.images = read_mnist_images(os.path.join(
            config.data_path, self.folder, self.files[self.which_set]['images']
        ))
        self.labels = read_mnist_labels(os.path.join(
            config.data_path, self.folder, self.files[self.which_set]['labels']
        ))
```
Now you can simply edit the `data_folder` or `files` attributes if they moved. For this HDF5 class, you could do the same thing. Maybe we should factor out the path joining though, maybe into a mixin class with a `get_file_path` method that does something like:

```python
class FileDataset(object):
    def get_file_path(self, file):
        folder = getattr(self, 'folder', '')
        return os.path.join(config.data_path, folder, file)
```